javascript/ql/src/Declarations/UnreachableMethodOverloads.ql

```diff
@@ -0,0 +1,144 @@
+  result != sig
+}
+
+/**
+ * Gets the kind of the member-declaration. Either "static" or "instance".
+ */
+string getKind(MemberDeclaration m) {
+  if m.isStatic() then result = "static" else result = "instance"
+}
+
+/**
+ * A call-signature that originates from a MethodSignature in the AST.
+ */
+private class MethodCallSig extends CallSignatureType {
+    string name;
+    
+    MethodCallSig() {
+        exists(MethodSignature sig | 
+            this = sig.getBody().getCallSignature() and
+            name = sig.getName()
+        )
+    }
+    
+    /**
+     * Gets the name of any member that has this signature.
+     */
+    string getName() {
+        result = name
+    }
+}
+
+/**
+ * Holds if the two call call signatures could be overloads of each other and have the same parameter types.
+ */
+predicate matchingCallSignature(MethodCallSig method, MethodCallSig other) {
+  method.getName() = other.getName() and 
+  
+  method.getNumOptionalParameter() = other.getNumOptionalParameter() and
+  method.getNumParameter() = other.getNumParameter() and
+  method.getNumRequiredParameter() = other.getNumRequiredParameter() and
+  // purposely not looking at number of type arguments. 
+  
+  method.getKind() = other.getKind() and 
+  
+  
+  forall(int i | i in [0 .. -1 + method.getNumParameter()] | 
+    method.getParameter(i) = other.getParameter(i) // This is sometimes imprecise, so it is still a good idea to compare type annotations.
+  ) and 
+  
+  // shared type parameters are equal. 
+  forall(int i | i in [0 .. -1 + min(int num | num = method.getNumTypeParameter() or num = other.getNumTypeParameter())] |
+    method.getTypeParameterBound(i) = other.getTypeParameterBound(i)
+  )
+}
+
+/**
+ * Gets which overload index the MethodSignature has among the overloads of the same name.
+ */
+int getOverloadIndex(MethodSignature sig) {
+    sig.getDeclaringType().getMethodOverload(sig.getName(), result) = sig
+}
+
+/**
+ * Holds if the two method-signatures are overloads of each other and have the same parameter types.
+ */
+predicate signaturesMatch(MethodSignature method, MethodSignature other) {
+  // declared in the same interface/class.
+  method.getDeclaringType() = other.getDeclaringType() and
+  // same static modifier.
+  getKind(method) = getKind(other) and
+
+  // same name.
+  method.getName() = other.getName() and
+
+  // same number of parameters.
+  method.getBody().getNumParameter() = other.getBody().getNumParameter() and
+  
+  // The types are compared in matchingCallSignature. This is sanity-check that the textual representation of the type-annotations are somewhat similar.
+  forall(int i | i in [0 .. -1 + method.getBody().getNumParameter()] |
+    getParameterTypeAnnotation(method, i) = getParameterTypeAnnotation(other, i)   
+  ) and
+  
+  matchingCallSignature(method.getBody().getCallSignature(), other.getBody().getCallSignature())
+}
+
+ClassOrInterface getSuperClass(ClassOrInterface base) {
```

Comment:
This sounds pretty general purpose; what's the reason for not moving it into the standard library?

Reply:
Because I didn't consider it.  

I moved it into `Classes.qll`

@asger-semmle have a look? da42577cfbf46df9ad91edfb0fd5bd6fb06935ff